### 3 - 8 -2020 題目

> 函数中的 return 语句一定能够得到执行。这种说法正确吗？
>
> A 正确
>
> B 错误

答案：**B 錯誤**

## 解析思路

在函數中只要加進了 If 等等的條件句，並把 return 語句放進去，那麽只要條件達不到一定要求，return 語句就不能被執行。

假設我們自己寫一個以下的程序碼：

```python
print("If a number is bigger than 3, it's added by 2. If not, the system returns error message.","\n")

def checkAndAdd(num):
    global addedNum

    if num <= 3:
        return "error"    # 返回 error 錯誤信息
    else:
        num += 2     # 為該數字加 2
        print(num)

print(checkAndAdd(2) + "\n")     # 當數字為 2 時
print(checkAndAdd(6))    # 當數字為 6 時
```

輸出結果：

```
If a number is bigger than 3, it's added by 2. If not, the system returns error message. 

error

8
None
```

以上示範的是當一個數字大於 3，該數字便會加 2，並打印出來。如果該數字小於或等於 3，系統便會返回「error」錯誤信息。

根據上面的結果，你能看到當數字為 2 時，即小於 3，return “error” 語句是得到執行的，而當數字為 6 時，return “error” 語句便跳過執行了，然後該數字就加上 2 和打印出來了，最後還接著打印了 None 值。

這裏再說明一下為何出現了一個 None 值，根據 Python 官方文檔 [Built-in Constants](https://docs.python.org/3/library/constants.html?highlight=constants#None) 和 [The standard type hierarchy](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy) 章節。

> None
>
> The sole value of the type `NoneType`. `None` is frequently used to represent the absence of a value, as when default arguments are not passed to a function. Assignments to `None` are illegal and raise a [`SyntaxError`](https://docs.python.org/3/library/exceptions.html#SyntaxError).

> None
>
> This type has a single value. There is a single object with this value. This object is accessed through the built-in name `None`. It is used to signify the absence of a value in many situations, e.g., it is returned from functions that don’t explicitly return anything. Its truth value is false.

None 值是 NoneType 對象的值，表示沒有任何值的意思。例如這個值會出現在當一個函數沒有任何返回值時，就會自動返回一個 None 值。